core/headerchain.go

```diff
@@ -413,6 +413,56 @@ func (hc *HeaderChain) SetCurrentHeader(head *types.Header) {
 // each header is deleted.
 type DeleteCallback func(common.Hash)
 
+func (hc *HeaderChain) PurgeAbove(n uint64, scanHeight uint64, delFn DeleteCallback) {
+
+	glog.V(logger.Warn).Infof("Purging block data above #%d", n)
+
+	// Set up logging for block purge progress.
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+	go func() {
+		for range ticker.C {
+			glog.V(logger.Warn).Infof("Removed data through #%d", n)
+		}
+	}()
+
+	// Search for any existing headers within n+reach
+	scanAfter := func(n uint64, reach uint64) bool {
+		var anyExistingFound bool
+		defer func() {
+			glog.V(logger.Debug).Infof("scanAfter n=%d bound=%d found=%v", n, n+reach, anyExistingFound)
+		}()
+		for i := n; i < n + reach; i++ {
+			if hc.GetHeaderByNumber(i) != nil {
+				anyExistingFound = true
+				return anyExistingFound
+			}
+		}
+		return anyExistingFound
+	}
+
+	// As long as there are any existing headers above the last check upper bound scope,
+	// remove them.
+	for scanAfter(n, scanHeight) {
```

Comment:
Instead of having two 3 loops (`for scanAfter`, loop in `scanAfter`, loop for removing), we could have single loop, and just keep track of last found header. And use distance between currently tested block number and last found as a predicate to exit loop.